lib/MergeStream.js

```diff
@@ -0,0 +1,280 @@
+var Readable = require('stream').Readable,
+    Transform = require('stream').Transform,
+    util = require('util'),
+    moment = require('moment'),
+    through2 = require('through2');
+
+/** A merger returns connections from different endpoints while maintaining departure time ordered. */
+var MergeStream = function(connectionsStreams, departureTimeSync) {
+  Readable.call(this, {objectMode: true});
+  var self = this;
+  this._connectionsStreams = [];
+  this._amountStreams = 0;
+  this._connectionsStreamIndex = {}; // map connectionsStream name on index
+  this._queues = []; // Every queue contains ordered connections
+  this._departureTime = departureTimeSync; // Keeps track of departureTime that we return connections from
+  this._maxInQueues = departureTimeSync; // Maximum departure time in queue
+  this._amountConnectionsInQueues = {}; // Counts amount of connections inside queues per stream
+  this._streamActive = {}; // holds bool whether a stream is being processed or not
+
+  for (var k=0; k<connectionsStreams.length; k++) {
+    this.addConnectionsStream(connectionsStreams[k]);
+  }
+};
+
+util.inherits(MergeStream, Readable);
+
+/**
+ * Adds a connections stream to the merger
+ * connectionsStream is array of stream name and stream itself
+ */
+MergeStream.prototype.addConnectionsStream = function (connectionsStream) {
+  var self = this;
+  var index = this._connectionsStreams.length;
+  var streamName = connectionsStream[0];
+  this._connectionsStreamIndex[streamName] = index; // map name - index stream
+  this._amountConnectionsInQueues[streamName] = 0;
+  this._amountStreams++;
+  this._streamActive[streamName] = true;
+
+  this._connectionsStreams[index] = connectionsStream[1]
+                                    .pipe(new AddStreamNameTransformer(streamName))
+                                    .pipe(through2.obj(function (connection, enc, done) {
+                                      // Don't process departed connection
+                                      if ((connection['@context'] || connection['departureTime'] >= self._departureTime) && self._streamActive[streamName]) {
+                                        this.push(connection);
```

Comment:
I'd implement this as a new transformer in a new file, so that we can also test how "heavy" this is by turning it off when optimizing this code. I really want the option to turn it off at some point in the future as I believe that you can be sure at this point that no `@context` is passed anymore (indeed: the context should already be processed here). On top of that: even if you still want to make sure there's not `@context`, didn't you forget a "not" (`!`) over here? And on top of that - bis -, I believe making sure that no departed connections are processed is something the fetcher code should do, not this code. Giving the option to turn it off by putting it in a different file seems appropriate.

Furthermore, I don't like using through2 for this: it puts an extra dependency on this repo. Putting this in in a different file is better code practise.